pkg/operator/controller/gatewayapi/crds.go

```diff
@@ -0,0 +1,128 @@
+package gatewayapi
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+
+	"github.com/openshift/cluster-ingress-operator/pkg/manifests"
+
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+)
+
+// managedCRDs is a list of CRDs that this controller manages.
+var managedCRDs = []*apiextensionsv1.CustomResourceDefinition{
+	manifests.GatewayClassCRD(),
+	manifests.GatewayCRD(),
+	manifests.HTTPRouteCRD(),
+}
+
+// ensureCRD attempts to ensure that the specified CRD exists and returns a
+// Boolean indicating whether it exists, the CRD if it does exist, and an error
+// value.
+func (r *reconciler) ensureCRD(ctx context.Context, desired *apiextensionsv1.CustomResourceDefinition) (bool, *apiextensionsv1.CustomResourceDefinition, error) {
+	name := types.NamespacedName{
+		Namespace: desired.Namespace,
+		Name:      desired.Name,
+	}
+	have, current, err := r.currentCRD(ctx, name)
+	if err != nil && !errors.IsNotFound(err) {
+		return have, current, err
+	}
+
+	switch {
+	case !have:
+		if err := r.createCRD(ctx, desired); err != nil {
+			return false, nil, err
+		}
+		return r.currentCRD(ctx, name)
+	case have:
+		if updated, err := r.updateCRD(ctx, current, desired); err != nil {
+			return have, current, err
+		} else if updated {
+			return r.currentCRD(ctx, name)
+		}
+	}
+	return have, current, nil
+}
+
+// ensureGatewayAPICRDs ensures the managed Gateway API CRDs are created and
+// returns an error value.  For now, the managed CRDs are the Gateway and
+// HTTPRoute CRDs.
+func (r *reconciler) ensureGatewayAPICRDs(ctx context.Context) error {
+	var errs []error
+	for i := range managedCRDs {
+		// The controller-runtime client mutates its argument, so give
+		// it a copy of the CRD rather than the original.
+		crd := managedCRDs[i].DeepCopy()
+		_, _, err := r.ensureCRD(ctx, crd)
+		errs = append(errs, err)
+	}
+	return utilerrors.NewAggregate(errs)
+}
+
+// currentCRD returns a Boolean indicating whether an CRD
+// exists for the IngressController with the given name, as well as the
+// CRD if it does exist and an error value.
+func (r *reconciler) currentCRD(ctx context.Context, name types.NamespacedName) (bool, *apiextensionsv1.CustomResourceDefinition, error) {
```

Comment:
I wonder if we should be storing and checking versions on CRDs?  And in the feature gate name.  When we move from v1beta1 to v1, will there be a chicken and egg upgrade problem?

Reply:
Yeah, I'll add a to-do for updating CRDs.  I'm not sure what you mean about the feature gate name.  Keep in mind that we don't support upgrades in dev preview.

Reply:
I added a to-do comment in https://github.com/openshift/cluster-ingress-operator/compare/9742e1ad86e7114b21eb78fde9be83caccc596b3..d4ac3988e55abd6c35f865e03366a20245ad4256.

Reply:
I removed the to-do comment; see https://github.com/openshift/cluster-ingress-operator/pull/890#discussion_r1113254640.